""" ИНФО
|-------------------------------------------------------------------------------|
|   lesson7.py - файл скрипта на языке python, хранится в текстовом виде,       |
|   открывается в любом текстовом редакторе (блокнот, notepad++)                |
|-------------------------------------------------------------------------------|
|   может быть выполнен в PyCharm с помощью сочетания клавиш                    |
|   Ctrl + Shift + F10                                                          |
|-------------------------------------------------------------------------------|
|   может быть выполнен из консоли командой                                     |
|   python lesson7.py                                                           |
|   при условии, что терминал консоли открыт в той же папке, что и сам скрипт   |
|-------------------------------------------------------------------------------|
"""


# подробный пример простой функции
def main():
    """ Функция main()
        def - ключевое слово для объявления функции (define/definition)
        main - имя функции
        print('hello world') - тело функции
    """
    print('hello world')


# еще один пример простой функции
def print_greeting():
    """ Функция print_greeting()
        по результату ввода пользователя печатает приветствие
    """
    name = input('Как Вас зовут?\n')
    print(f'Здравствуйте, {name}!')


# пример функции с двумя аргументами
def count_summ(a, b):
    """ Считает сумму двух чисел

    Аргументы:
        a : первое слагаемое
        b : второе слагаемое

    Возвращаемое значение:
        сумма a и b
    """
    summ = a + b
    return summ


# пример функции с бесконечным циклом
def endless_function():
    print('Остановите меня! (Ctrl + C)')
    while True:
        pass


# пример функции, которая вызывает исключение
def bad_function():
    raise ValueError('Произошло что-то непонятное...')


# пример функции, обрабатывающей различные исключения
def exceptions_example():
    try:
        print('\nПробуем выполнить "count_summ(2)"')
        count_summ(2)
    except TypeError as err:
        print(err)
    
    try:
        print('\nПробуем выполнить "greet_printing()"')
        greet_printing()
    except NameError as err:
        print(err)

    try:
        print('\nПробуем выполнить "endless_function()"')
        endless_function()
    except KeyboardInterrupt:
        print('Бесконечный цикл был прерван извне')

    try:
        print('\nПробуем выполнить "bad_function()"')
        bad_function()
    except ValueError as err:
        print(err)
    
    print('Все исключения были успешно обработаны!')


# начало работы скрипта
# специальная конструкция для вызова фнукции по-умолчанию
if __name__ == '__main__':
    exceptions_example()
